numba/np/ufunc/deviceufunc.py

```diff
@@ -732,110 +724,164 @@ def _broadcast_scalar_input(self, ary, shape):
+                convert = self.as_device_array
+            else:
+                convert = np.asarray
+
+            return convert(a)
+
+        normalized_inputs = [normalize_input(a) for a in args]
+        self.inputs = normalized_inputs[:nin]
+
         # Check if there are extra arguments for outputs.
-        unused_inputs = inputs[nin:]
+        unused_inputs = normalized_inputs[nin:]
         if unused_inputs:
-            if self.output is not None:
-                raise ValueError("cannot specify 'out' as both a positional "
-                                 "and keyword argument")
-            else:
-                [self.output] = unused_inputs
+            self.outputs = unused_inputs
 
     def adjust_input_types(self, indtypes):
         """
         Attempt to cast the inputs to the required types if necessary
-        and if they are not device array.
+        and if they are not device arrays.
 
-        Side effect: Only affects the element of `norm_inputs` that requires
+        Side effect: Only affects the elements of `inputs` that require
         a type cast.
         """
-        for i, (ity, val) in enumerate(zip(indtypes, self.norm_inputs)):
+        for i, (ity, val) in enumerate(zip(indtypes, self.inputs)):
             if ity != val.dtype:
                 if not hasattr(val, 'astype'):
                     msg = ("compatible signature is possible by casting but "
                            "{0} does not support .astype()").format(type(val))
                     raise TypeError(msg)
                 # Cast types
-                self.norm_inputs[i] = val.astype(ity)
+                self.inputs[i] = val.astype(ity)
 
-    def allocate_outputs(self, schedule, outdtype):
-        # allocate output
-        if self._need_device_conversion or self.output is None:
-            retval = self.device_array(shape=schedule.output_shapes[0],
-                                       dtype=outdtype)
-        else:
-            retval = self.output
-        self.kernel_returnvalue = retval
-
-    def prepare_kernel_parameters(self):
-        params = []
-        for inp, isdev in zip(self.norm_inputs, self._is_device_array):
-            if isdev:
-                params.append(inp)
-            else:
-                params.append(self.to_device(inp))
-        assert all(self.is_device_array(a) for a in params)
-        self.kernel_parameters = params
-
-    def post_process_result(self):
-        if self._need_device_conversion:
-            out = self.to_host(self.kernel_returnvalue, self.output)
-        elif self.output is None:
-            out = self.kernel_returnvalue
-        else:
-            out = self.output
-        return out
+    def prepare_outputs(self, schedule, outdtypes):
+        """
+        Returns an array of output parameters that all reside on the target device.
 
-    def prepare_inputs(self):
-        pass
+        Outputs that were passed-in to the GUFunc are used if they reside on the
```

Comment:
I'm not sure this is clearer, because it's not immediately obvious to me whether "arguments" refers to formal or actual arguments, hence my use of "passed-in" to denote actual arguments. (I also don't want to write "actual arguments" because that also could be ambiguous in this context)

Reply:
Ah thanks, I see what you mean, let's leave it, "passed-in" whilst a little awkward provides the necessary context.